src/python/pants/pantsd/service/scheduler_service.py

```diff
@@ -191,89 +189,66 @@ def _process_event_queue(self):
-            self._logger.debug(
-                "graph len was {}, waiting for initial watchman event".format(graph_len)
-            )
+            self._logger.debug(f"graph len was {graph_len}, waiting for initial watchman event")
             self._watchman_is_running.wait()
+
+        global_options = options.for_global_scope()
         build_id = RunTracker.global_instance().run_id
-        v2_ui = options.for_global_scope().get("v2_ui", False)
+        v2_ui = global_options.get("v2_ui", False)
         zipkin_trace_v2 = options.for_scope("reporting").zipkin_trace_v2
-        session = self._graph_helper.new_session(zipkin_trace_v2, build_id, v2_ui)
-
-        if options.for_global_scope().get("loop", False):
-            fn = self._loop
-        else:
-            fn = self._body
+        return self._graph_helper.new_session(zipkin_trace_v2, build_id, v2_ui)
 
-        specs, exit_code = fn(session, options, options_bootstrapper)
-        return session, specs, exit_code
-
-    def _loop(
+    def graph_run_v2(
         self,
         session: LegacyGraphSession,
+        specs: Specs,
         options: Options,
         options_bootstrapper: OptionsBootstrapper,
-    ) -> Tuple[Specs, int]:
-        # TODO: See https://github.com/pantsbuild/pants/issues/6288 regarding Ctrl+C handling.
-        iterations = options.for_global_scope().loop_max
-        specs = None
-        exit_code = PANTS_SUCCEEDED_EXIT_CODE
-        while iterations and not self._state.is_terminating:
-            try:
-                specs, exit_code = self._body(session, options, options_bootstrapper)
-            except session.scheduler_session.execution_error_type as e:
-                # Render retryable exceptions raised by the Scheduler.
-                print(e, file=sys.stderr)
-
-            iterations -= 1
-            while (
-                iterations
-                and not self._state.is_terminating
-                and not self._loop_condition.wait(timeout=1)
-            ):
-                continue
-        return cast(Specs, specs), exit_code
+    ) -> int:
+        """Perform an entire v2 run.
+
+        The exit_code in the return indicates whether any issue was encountered.
+        """
+
+        global_options = options.for_global_scope()
+        perform_loop = global_options.get("loop", False)
+        v2 = global_options.v2
+
+        if perform_loop:
+            # TODO: See https://github.com/pantsbuild/pants/issues/6288 regarding Ctrl+C handling.
+            loop_max = global_options.loop_max
+            exit_code = PANTS_SUCCEEDED_EXIT_CODE
+            for _ in range(loop_max):
+                if self._state.is_terminating or self._loop_condition.wait(timeout=1):
+                    break
+                try:
+                    exit_code = self._body(session, options, options_bootstrapper, specs, v2)
+                except session.scheduler_session.execution_error_type as e:
+                    self._logger.warning(e)
```

Comment:
For a refactoring change, this part feels rather large. For example, we're doing `self._loop_condition.wait(timeout=1)` as a condition in our `if` instead of checking `iterations` as in the previous version. Is it possible to refactor this method to look more like the original code, while keeping the rest of the PR the same?